library/std/src/os/horizon/thread.rs

```diff
@@ -0,0 +1,88 @@
+//! Nintendo 3DS-specific extensions to primitives in the [`std::thread`] module.
+//!
+//! All 3DS models have at least two CPU cores available to spawn threads on:
+//! The application core (appcore) and the system core (syscore). The New 3DS
+//! has an additional two cores, the first of which can also run user-created
+//! threads.
+//!
+//! Threads spawned on the appcore are cooperative rather than preemptive. This
+//! means that threads must explicitly yield control to other threads (whether
+//! via synchronization primitives or explicit calls to `yield_now`) when they
+//! are not actively performing work. Failure to do so may result in control
+//! flow being stuck in an inactive thread while the other threads are powerless
+//! to continue their work.  
+//!
+//! However, it is possible to spawn one fully preemptive thread on the syscore
+//! by using a service call (https://www.3dbrew.org/wiki/APT:SetApplicationCpuTimeLimit)
+//! to reserve a slice of time for a thread to run. Attempting to run more than
+//! one thread at a time on the syscore will result in an error.
+//!
+//! [`std::thread`]: crate::thread
+
+#![unstable(feature = "horizon_thread_ext", issue = "none")]
+
+/// Extensions on [`std::thread::Builder`] for the Nintendo 3DS.
+///
+/// [`std::thread::Builder`]: crate::thread::Builder
+pub trait ThreadBuilderExt: Sized {
+    /// Sets the priority level for the new thread.
+    ///
+    /// Low values gives the thread higher priority. For userland apps, this has
+    /// to be within the range of 0x18 to 0x3F inclusive. The main thread
+    /// usually has a priority of 0x30, but not always.
+    fn priority(mut self, priority: i32) -> Self;
+
+    /// Sets the ID of the processor the thread should be run on. Threads on the
+    /// 3DS are only preemptive if they are on the system core. Otherwise they
+    /// are cooperative (must yield to let other threads run).
+    ///
+    /// Processor IDs are labeled starting from 0. On Old3DS it must be <2, and
+    /// on New3DS it must be <4. Pass -1 to execute the thread on all CPUs and
+    /// -2 to execute the thread on the default CPU (set in the application's
+    /// Exheader).
+    ///
+    /// * Processor #0 is the application core. It is always possible to create
+    ///   a thread on this core.
+    /// * Processor #1 is the system core. If the CPU time limit is set, it is
+    ///   possible to create a single thread on this core.
+    /// * Processor #2 is New3DS exclusive. Normal applications can create
+    ///   threads on this core if the exheader kernel flags bitmask has 0x2000
+    ///   set.
+    /// * Processor #3 is New3DS exclusive. Normal applications cannot create
+    ///   threads on this core.
+    fn ideal_processor(mut self, ideal_processor: i32) -> Self;
```

Comment:
Naming wise, should this be just `processor` or still `ideal_processor`? Using `affinity` might be confusing since there's a separate affinity mask, and `ideal_processor` seems a little long/wordy.

This naming should probably be the same across std and the nonstandard libc functions.

Reply:
I think `ideal_processor` is fine, since there is some reference to this wording in `libctru` already, but it could also be `processor_id` to match [the libctru argument](https://libctru.devkitpro.org/svc_8h.html#a3f62a8f1822c6a78fab2e2cb03c46893).

Reply:
I think I'll switch to `processor_id` to make things a little less... exotic.

Reply:
Opened https://github.com/Meziu/libc/pull/10 and https://github.com/Meziu/pthread-3ds/pull/11